### PR TITLE
WIP - GraphQL

### PR DESCRIPTION
**Pull request by RichardLitt**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,4 @@
 node_modules
+data
+
+package-lock.json
```
Comment (RichardLitt) [inline .gitignore]: Should be checked in.

File: package.json [hunk 2]
```diff
@@ -1,63 +1,53 @@
...
-    "standard": "^10.0.3",
-    "xo": "^0.18.2",
-    "semantic-release": "^8.0.0"
-  },
-  "coordinates": [
-    45.5231596,
-    -73.5870259
-  ]
+	"name": "name-your-contributors",
+	"version": "0.0.1",
```
Comment (RichardLitt) [inline package.json]: Leave it as 0.0.0-development. Using [semantic-release](https://github.com/semantic-release/semantic-release).

File: src/cli.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+`, {
+	alias: {
+		b: 'before',
+		a: 'after',
+		r: 'repo',
+		u: 'user'
+	}
+}]);
+
+const token = process.env.GITHUB_TOKEN;
```
Comment (RichardLitt) [inline src/cli.js]: Add a `t: 'token'` so you can specify this from your CLI.

Comment (RichardLitt) [reply src/cli.js]: Is a token necessary?

File: src/cli.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+if (cli.flags.u && cli.flags.r && token) {
+	graphql.executequery(token, queries.everything(cli.flags.r, cli.flags.u))
```
Comment (RichardLitt) [inline src/cli.js]: if no since or r, get all of the contributors forever for a repo.

File: src/graphql.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
+'use strict'
+
+const https = require('https');
+
+/** Escape strings to prevent injection attacks. Other types aren't an issue. */
+const escapeArgValue = val => {
+	if ('string' === typeof (val)) {
```
Comment (RichardLitt) [inline src/graphql.js]: Can be shortened: `return ('string' === typeof (val)) ? ... : val`

File: src/graphql.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		return '"' + encodeURIComponent(val) + '"';
+	} else {
+		return val;
+	}
+};
+
+const argsString = args => {
+	const keys = Object.keys(args);
+	if (keys.length === 0) {
+		return '';
```
Comment (RichardLitt) [inline src/graphql.js]: Can you lint according to `standard`?

File: src/queries.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	participantsQ
+];
+
+const everything = (repoName, ownerName) =>
+			node('repository', {name: repoName, owner: ownerName})
+			.addChild(node('pullRequests', {first: 100}, multiQs))
+			.addChild(node('issues', {first: 100}, multiQs));
+
+const timeFilter = (before = new Date(), after = new Date(0)) =>
+			data => data.filter(x => after <= x.createdAt <= before);
```
Comment (RichardLitt) [inline src/queries.js]: wtf is this syntax? :D Never seen `<=`

File: src/queries.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+const users = arr =>
+	arr.map(x => x.author)
+			// Get rid of null authors (weird GitHub problem)
+			// Deleted accounts maybe?
```
Comment (RichardLitt) [inline src/queries.js]: bingo. That's what causes this.

File: src/queries.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+			.filter(x => !(null === x || undefined === x))
+			.map(x => [x.login, x.url]);
+
+const flatten = arr => arr.reduce((acc, next) => acc.concat(next), []);
```
Comment (RichardLitt) [inline src/queries.js]: Doesn't lodash have this as a method? import it directly?

File: test/test.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+});
+
+test('With a constrained since and until', t => {
+	return fn('opensourcedesign', {
+		since: '2016-01-01T00:01:01Z',
+		until: '2016-01-04T20:00:00Z'
+	}).then(result => {
+		t.same(result, ['[@bnvk](//github.com/bnvk) (Brennan Novak)', '[@jdittrich](//github.com/jdittrich)']);
+	});
+});
```
Comment (RichardLitt) [inline test/test.js]: Can we test for each of the exported functions?

**Comment by RichardLitt** [comment]
Change folder it saves to to fullname: aka `data/RichardLitt/name-your-contributors`

File: src/cli.js (hunk 3 above)
Comment (tgetgood) [reply src/cli.js]: The v4 API requires a token for any requests.

File: src/graphql.js (hunk 6 above)
Comment (tgetgood) [reply src/graphql.js]: Standard is very opinionated, not configurable, and not standard. 

Of the 3 linters that we previously in use I chose xo because it was quick and configurable. We can set the linting flags as you like, but standard makes some assumptions that I would consider bad style, and the fact that you can't configure it was making it inordinately difficult.

File: src/queries.js (hunk 7 above)
Comment (tgetgood) [reply src/queries.js]: Less than or equal. 

That line was actually wrong as well, you can't say a < b < c in javascript, though it's  not an error, it's just undefined...

File: src/queries.js (hunk 9 above)
Comment (tgetgood) [reply src/queries.js]: My general philosophy is that if something takes less than a minute to implement, downloading, and learning a library is a negative time gain. It also makes the package size way bigger.

File: test/test.js (hunk 10 above)
Comment (tgetgood) [reply test/test.js]: yeah, tests have been totally redone.

File: src/graphql.js (hunk 5 above)
Comment (tgetgood) [reply src/graphql.js]: It certainly can. Would that make it any better though?

**Comment by tgetgood** [comment]
I've since gotten rid of the data directory. It's not needed. 

Otherwise the only thing left to do is deal with paginated requests (github only lets you get 100 entries of any given thing at a time). 

Everything should be in working order but only the first 100 issues get queried for a repo and the first 100 comments from each issue, etc.. This will be fixed shortly.

Also the CI server needs a token. The v4 API requires one.

**Comment by tgetgood** [comment]
An update on pagination: I've run into a couple of issues with the API. The error messages from the server say it's probably a problem on their end, so I've sent off an email to support. I'll post any updates.

**Comment by RichardLitt** [comment]
> Also the CI server needs a token. The v4 API requires one.

So each user will have to get their own? That's OK, I'm cool with that. 

Thanks for checking on pagination. 

What do you think about this little module, so far?

File: src/graphql.js (hunk 6 above)
Comment (RichardLitt) [reply src/graphql.js]: Yeah. I use standard because I know feross, and I'd rather use his stylguide then waste time on my own. Spending time figuring out my own opinion isn't where I want to be, either. XO is similar, AFAIK - the option seems arbitrary. You can always set ignores for issues with particular files or lines - that what I do sometimes to get around standard being persnickety.

File: src/graphql.js (hunk 5 above)
Comment (RichardLitt) [reply src/graphql.js]: Nah.

File: src/queries.js (hunk 7 above)
Comment (RichardLitt) [reply src/queries.js]: Ah. Wow, hard to parse.

File: readme.md [hunk 11]
```diff
@@ -32,63 +42,82 @@ The environmental variable is also used by several of `name-your-contributor`'s
...
-nameYourContributors('ipfs', {
-  since: '2016-01-15T00:20:24Z'
-});
-//=> '[@RichardLitt](//github.com/RichardLitt) (Richard Littauer)'
+```js
+const nyc = require('name-your-contributors');
+
+nyc.nameYourContributors({
+	token: process.env.GITHUB_TOKEN,
+	user: 'RichardLitt,
```
Comment (RichardLitt) [inline readme.md]: needs closing '

File: src/cli.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	Examples
+		$ name-your-contributors -r ipfs -u ipfs --after=2016-01-15T00:20:24Z --before=2016-01-20T00:20:24Z
+
+		$ name-your-contributors -o ipfs -a 2017-01-01 > ipfs-contrib-2017.json
+`, {
+	alias: {
+		b: 'before',
+		a: 'after',
+		r: 'repo',
+		u: 'user'
```
Comment (RichardLitt) [inline src/cli.js]: Needs t: token to be set

**Comment by RichardLitt** [comment]
LMK when this is ready to review!

**Comment by tgetgood** [comment]
Ready to review. The main features are all in, the pagination code is somewhat boilerplaty, but since we only have a couple of queries it doesn't much matter. 

Careful with the `-o org-name` option since it can make a lot of requests. A single `-o wikimedia` will exhaust your hourly quota. The before and after flags don't effect this since the graphql API gives no way to filter in the query; the filtering all happens on the query response. 

As for CI, I don't know how you want to do it. Giving travisci a token in the source is just asking for it to get abused, and hitting a third party api in an automated test is boarderline spammy. I'd say just test against hard coded queries and dummy responses (to check the queries don't get mangled, and the post-processors still work). That will break if the API changes, but things will break in that event no matter what. What's your opinion here?

**Comment by RichardLitt** [comment]
Awesome. Boilerplate pagination is fine. 

Add a note about the orgname possible exhausting the quota to the README, if you could. 

I added a token to Travis, but it won't be seen in the source. Tests passed locally - let's see if they pass again. I wouldn't worry about hitting GitHub in the API, I think that's OK. Unless I'm committing a grievous sin, there? We could spoof the API, but not sure how to spoof a GraphQL api.

**Comment by RichardLitt** [comment]
Any chance we could easily get normal contributions in this, too? People who commit code, for instance. Or through GitHub's API.

Same for people who make reactions (:+1:).

**Comment by tgetgood** [comment]
Note added. I also removed the section of the readme regarding other apps using the same auth token, since I gather that those other apps will no longer be necessary. 

CI is still failing with auth errors. What's the token exported as?

It would be very straight forward to add more branches into the query.  Currently you have to match those with pagination, parsing, and aggregation, which as I said before is more boilerplate than I'd like. Easy, but repetitive. For just those two additions it's probably still worth it, but if we're going to want a lot of changes we'll eventually want a better graphql library. I couldn't find one that served our purpose very well: shopify's is the best as far as I can see, but it still doesn't have any capacity for automatic depaginating, filtering, or custom aggregations. I'm generally loath to write my own libraries for things but there seems to be a big gap in the space. The clients I saw all target browsers doing something relayish, whereas we're doing analytics.

Hopefully there's a better library that I missed, I didn't spend all that much time scouring.  Any ideas?

**Comment by RichardLitt** [comment]
> CI is still failing with auth errors. What's the token exported as?

Looks like Travis [doesn't allow this for forks](https://docs.travis-ci.com/user/pull-requests#Pull-Requests-and-Security-Restrictions). It runs locally just fine.

**Comment by RichardLitt** [comment]
> It would be very straight forward to add more branches into the query. Currently you have to match those with pagination, parsing, and aggregation, which as I said before is more boilerplate than I'd like. Easy, but repetitive. For just those two additions it's probably still worth it, but if we're going to want a lot of changes we'll eventually want a better graphql library. I couldn't find one that served our purpose very well: shopify's is the best as far as I can see, but it still doesn't have any capacity for automatic depaginating, filtering, or custom aggregations. I'm generally loath to write my own libraries for things but there seems to be a big gap in the space. The clients I saw all target browsers doing something relayish, whereas we're doing analytics.

> Hopefully there's a better library that I missed, I didn't spend all that much time scouring. Any ideas?

Can you open an issue for this? We can track it as we go. Right now, I would like to get commits too, for sure, although I guess I can write a thing to track `git authors` and merge them. (Maybe another issue for this?) 

Regarding a better library: hmm. We could write a better one, but I want to make sure it is used for more than just this repository. Let's just track it in an issue.
